File: Zigote.Http/Http.cs
using System.Collections.Immutable;
using System.Globalization;
using System.Text.Json.Serialization.Metadata;
using Zigote.Http.Frame;

namespace Zigote.Http;

/// <summary>
///     The entry point for building requests. Pure — <c>HttpRequest.Get(...)</c> returns a value,
///     and every transform below returns a new one. Nothing is sent until a
///     <see cref="HttpRunner" /> runs it, which is what makes a request loggable, cacheable and
///     testable before it exists on a wire.
/// </summary>
/// <remarks>
///     The design doc called this <c>Http</c>. It cannot be: a type named <c>Zigote.Http.Http</c>
///     is shadowed by the <c>Zigote.Http</c> namespace for every caller that itself lives under
///     <c>Zigote</c> — which is all of them. The frame queue moved to <see cref="HttpFrame" /> in
///     the same breath, and the two surfaces read better apart anyway.
/// </remarks>
public static class HttpRequest
{
    /// <summary>A GET for <paramref name="template" />, which may contain <c>{placeholders}</c>.</summary>
    public static HttpSpec Get(string template) => HttpSpec.For(HttpVerb.Get, template);

    /// <summary>A HEAD.</summary>
    public static HttpSpec Head(string template) => HttpSpec.For(HttpVerb.Head, template);

    /// <summary>A POST.</summary>
    public static HttpSpec Post(string template) => HttpSpec.For(HttpVerb.Post, template);

    /// <summary>A PUT.</summary>
    public static HttpSpec Put(string template) => HttpSpec.For(HttpVerb.Put, template);

    /// <summary>A PATCH.</summary>
    public static HttpSpec Patch(string template) => HttpSpec.For(HttpVerb.Patch, template);

    /// <summary>A DELETE.</summary>
    public static HttpSpec Delete(string template) => HttpSpec.For(HttpVerb.Delete, template);

}

/// <summary>
///     The frame loop's submit/poll queue, as a host-assigned provider: assigned at startup, faked
///     in tests, the same shape <c>Input</c> and <c>Audio</c> already use. Every method here
///     allocates nothing, so widgets and gameplay code can use them inside Measure→Layout→Paint.
/// </summary>
public static class HttpFrame
{
    /// <summary>The queue these methods forward to. Null until the host assigns one.</summary>
    public static FrameHttpQueue? Queue { get; set; }

    /// <summary>Queues a request from the frame path. Invalid handle when there is no queue or it is full.</summary>
    public static HttpHandle Submit(HttpSpec spec) => Queue?.Submit(spec) ?? default;

    /// <summary>Polls a submitted request. Allocates nothing.</summary>
    public static bool TryTake(HttpHandle handle, out HttpOutcome outcome)
    {
        var queue = Queue;
        if (queue is not null) return queue.TryTake(handle, out outcome);
        outcome = default;
        return false;
    }

    /// <summary>Asks a submitted request to stop.</summary>
    public static void Cancel(HttpHandle handle) => Queue?.Cancel(handle);

    /// <summary>Frees the slot behind a taken handle, and its body buffer with it.</summary>
    public static void Release(HttpHandle handle) => Queue?.Release(handle);
}

/// <summary>
///     The fluent surface: pure transforms over a <see cref="HttpSpec" />, one per thing a request
///     can carry. Named <c>With*</c> where the spec already has a property of that name, so the
///     builder and the data never shadow each other.
/// </summary>
public static class HttpSpecExtensions
{
    /// <summary>Binds a <c>{placeholder}</c> in the route template. The value is percent-encoded at render time.</summary>
    public static HttpSpec Route(this HttpSpec spec, string name, string value) =>
        spec with { Path = spec.Path.Bind(name, value) };

    /// <summary>Binds a <c>{placeholder}</c> from anything formattable, invariantly.</summary>
    public static HttpSpec Route<T>(this HttpSpec spec, string name, T value) where T : IFormattable =>
        spec.Route(name, value.ToString(null, CultureInfo.InvariantCulture));

    /// <summary>Appends a query-string pair.</summary>
    public static HttpSpec WithQuery(this HttpSpec spec, string name, string value) =>
        spec with { Query = Safe(spec.Query).Add(new QueryParam(name, value)) };

    /// <summary>Appends a query-string pair from anything formattable, invariantly.</summary>
    public static HttpSpec WithQuery<T>(this HttpSpec spec, string name, T value) where T : IFormattable =>
        spec.WithQuery(name, value.ToString(null, CultureInfo.InvariantCulture));

    /// <summary>Appends a query-string pair only when <paramref name="value" /> is not null. For optional filters.</summary>
    public static HttpSpec WithQueryIf(this HttpSpec spec, string name, string? value) =>
        value is null ? spec : spec.WithQuery(name, value);

    /// <summary>Adds a request header.</summary>
    public static HttpSpec WithHeader(this HttpSpec spec, string name, string value) =>
        spec with { Headers = Safe(spec.Headers).Add(new HeaderPair(name, value)) };

    /// <summary>Sets the body.</summary>
    public static HttpSpec WithBody(this HttpSpec spec, HttpBody body) => spec with { Body = body };

    /// <summary>Sets a JSON body, serialized now with a source-generated contract.</summary>
    public static HttpSpec WithJson<T>(this HttpSpec spec, T value, JsonTypeInfo<T> typeInfo) =>
        spec with { Body = HttpBody.Json(value, typeInfo) };

    /// <summary>Sets a form body.</summary>
    public static HttpSpec WithForm(this HttpSpec spec, params ReadOnlySpan<QueryParam> fields) =>
        spec with { Body = HttpBody.Form([..fields]) };

    /// <summary>
    ///     Sets a <c>multipart/form-data</c> body. Build parts with
    ///     <see cref="MultipartPart.Field" /> and <see cref="MultipartPart.File(string, string, ReadOnlyMemory{byte}, string)" />;
    ///     a request whose parts are all replayable stays retryable, one stream part makes it one-shot.
    /// </summary>
    public static HttpSpec WithMultipart(this HttpSpec spec, params ReadOnlySpan<MultipartPart> parts) =>
        spec with { Body = HttpBody.Multipart(parts) };

    /// <summary>Sets the budget for the whole call — retries and revalidation included.</summary>
    public static HttpSpec Deadline(this HttpSpec spec, TimeSpan budget) =>
        spec with { Policy = spec.Policy with { Deadline = budget } };

    /// <summary>Sets retry and backoff.</summary>
    public static HttpSpec Retry(this HttpSpec spec, RetryPolicy retry) =>
        spec with { Policy = spec.Policy with { Retry = retry } };

    /// <summary>Turns retry off for this request.</summary>
    public static HttpSpec NoRetry(this HttpSpec spec) => spec.Retry(RetryPolicy.None);

    /// <summary>Sets how this request treats the cache.</summary>
    public static HttpSpec Cache(this HttpSpec spec, CacheMode mode) =>
        spec with { Policy = spec.Policy with { Cache = mode } };

    /// <summary>Skips the cache in both directions.</summary>
    public static HttpSpec NoCache(this HttpSpec spec) => spec.Cache(CacheMode.Bypass);

    /// <summary>Declares that repeating this request is safe, so a non-idempotent verb may be retried.</summary>
    public static HttpSpec Idempotent(this HttpSpec spec) =>
        spec with { Policy = spec.Policy with { Idempotent = true } };

    /// <summary>Sends this request without an <c>Authorization</c> header.</summary>
    public static HttpSpec Anonymous(this HttpSpec spec) =>
        spec with { Policy = spec.Policy with { Anonymous = true } };

    /// <summary>Hands the body back as an open stream instead of buffering it.</summary>
    public static HttpSpec Streaming(this HttpSpec spec) =>
        spec with { Policy = spec.Policy with { Streaming = true } };

    /// <summary>Reports transfer progress for each attempt.</summary>
    public static HttpSpec Progress(this HttpSpec spec, IProgress<HttpProgress> progress) =>
        spec with { Policy = spec.Policy with { Progress = progress } };

    /// <summary>Replaces the whole policy.</summary>
    public static HttpSpec WithPolicy(this HttpSpec spec, RequestPolicy policy) => spec with { Policy = policy };

    private static ImmutableArray<T> Safe<T>(ImmutableArray<T> array) =>
        array.IsDefault ? ImmutableArray<T>.Empty : array;
}

File: Zigote.Http.FSharp/Http.fs
namespace Zigote.Http.FSharp

open System
open System.Net
open System.Text.Json.Serialization.Metadata
open Zigote.Http

/// <summary>
/// The F# surface: a request is a value, transforms are pure functions, and running one is a
/// separate, explicit step. Every function here takes the spec last, so they compose with |&gt;.
/// </summary>
/// <remarks>
/// <para>
/// This is not a wrapper around a C# API — <c>HttpSpec</c>, <c>HttpError</c> and
/// <c>HttpRunner</c> are the same types the C# side uses. What F# adds is the pipeline shape
/// and <c>Result</c>: <c>HttpResult&lt;'T&gt;</c> converts to <c>Result&lt;'T, HttpError&gt;</c>
/// with no adapter and no boxing, and <c>HttpError</c>'s abstract-record hierarchy
/// pattern-matches natively.
/// </para>
/// <para>
/// An <c>http { ... }</c> computation expression is deliberately absent: the pipeline functions
/// already cover the ergonomics, and a CE built before the surface settles is a CE that has to
/// be unbuilt.
/// </para>
/// </remarks>
[<RequireQualifiedAccess>]
module Http =

    /// A GET for a route template, which may contain {placeholders}.
    let get (template: string) = HttpSpec.For(HttpVerb.Get, template)

    /// A HEAD.
    let head (template: string) = HttpSpec.For(HttpVerb.Head, template)

    /// A POST.
    let post (template: string) = HttpSpec.For(HttpVerb.Post, template)

    /// A PUT.
    let put (template: string) = HttpSpec.For(HttpVerb.Put, template)

    /// A PATCH.
    let patch (template: string) = HttpSpec.For(HttpVerb.Patch, template)

    /// A DELETE.
    let delete (template: string) = HttpSpec.For(HttpVerb.Delete, template)

    /// Binds a {placeholder} in the route template.
    let route (name: string) (value: string) (spec: HttpSpec) = spec.Route(name, value)

    /// Appends a query-string pair. The value is formatted invariantly.
    let query (name: string) (value: obj) (spec: HttpSpec) =
        spec.WithQuery(name, Convert.ToString(value, Globalization.CultureInfo.InvariantCulture))

    /// Adds a request header.
    let header (name: string) (value: string) (spec: HttpSpec) = spec.WithHeader(name, value)

    /// Sets the body.
    let body (b: HttpBody) (spec: HttpSpec) = spec.WithBody b

    /// Sets a JSON body, serialized now with a source-generated contract.
    let jsonBody (typeInfo: JsonTypeInfo<'T>) (value: 'T) (spec: HttpSpec) = spec.WithJson(value, typeInfo)

    /// Sets a multipart/form-data body from fields and files (see MultipartPart.Field / .File).
    let multipart (parts: MultipartPart list) (spec: HttpSpec) =
        spec.WithBody(HttpBody.MultipartBody(Collections.Immutable.ImmutableArray.CreateRange parts))

    /// Reports transfer progress — download always, upload when the request has a body.
    let progress (sink: IProgress<HttpProgress>) (spec: HttpSpec) = spec.Progress sink

    /// Sets the budget for the whole call — retries and revalidation included.
    let deadline (budget: TimeSpan) (spec: HttpSpec) = spec.Deadline budget

    /// Sets how this request treats the cache.
    let cache (mode: CacheMode) (spec: HttpSpec) = spec.Cache mode

    /// Sets retry and backoff.
    let retry (policy: RetryPolicy) (spec: HttpSpec) = spec.Retry policy

    /// Declares that repeating this request is safe.
    let idempotent (spec: HttpSpec) = spec.Idempotent()

    /// Sends this request without an Authorization header.
    let anonymous (spec: HttpSpec) = spec.Anonymous()

    /// An HttpResult as an F# Result. No adapter, no boxing — the struct is already this shape.
    let inline toResult (result: HttpResult<'T>) : Result<'T, HttpError> =
        if result.IsOk then Ok result.Value else Error result.Error

    /// "That status is an answer, not an error" — a 404 becomes the fallback, everything else passes through.
    let recover (code: Net.HttpStatusCode) (fallback: 'T) (result: Result<'T, HttpError>) =
        match result with
        | Error(:? HttpError.Status as s) when s.Code = code -> Ok fallback
        | other -> other

    /// Runs a spec and hands back the response. The caller disposes it.
    let send (spec: HttpSpec) (runner: HttpRunner) : Async<Result<HttpResponse, HttpError>> =
        async {
            let! ct = Async.CancellationToken
            let! result = runner.SendAsync(spec, ct).AsTask() |> Async.AwaitTask
            return toResult result
        }

    /// Runs a spec and decodes a 2xx body as JSON.
    let json (typeInfo: JsonTypeInfo<'T>) (spec: HttpSpec) (runner: HttpRunner) : Async<Result<'T, HttpError>> =
        async {
            let! ct = Async.CancellationToken
            let! result = runner.JsonAsync(spec, typeInfo, ct).AsTask() |> Async.AwaitTask
            return toResult result
        }

    /// Runs a spec and hands back the 2xx body bytes.
    let bytes (spec: HttpSpec) (runner: HttpRunner) : Async<Result<byte[], HttpError>> =
        async {
            let! ct = Async.CancellationToken
            let! result = runner.BytesAsync(spec, ct).AsTask() |> Async.AwaitTask
            return toResult result
        }

    /// Runs a spec and hands back the 2xx body as UTF-8 text.
    let text (spec: HttpSpec) (runner: HttpRunner) : Async<Result<string, HttpError>> =
        async {
            let! ct = Async.CancellationToken
            let! result = runner.TextAsync(spec, ct).AsTask() |> Async.AwaitTask
            return toResult result
        }

    // ── task-shaped variants ─────────────────────────────────────────────────
    // The same calls for code living in a `task { }` — modern F# services and anything talking to
    // C# APIs. Cancellation is explicit here because task, unlike async, carries no ambient token.

    /// `send`, task-shaped.
    let sendTask (spec: HttpSpec) (ct: Threading.CancellationToken) (runner: HttpRunner) =
        task {
            let! result = runner.SendAsync(spec, ct)
            return toResult result
        }

    /// `json`, task-shaped.
    let jsonTask (typeInfo: JsonTypeInfo<'T>) (spec: HttpSpec) (ct: Threading.CancellationToken) (runner: HttpRunner) =
        task {
            let! result = runner.JsonAsync(spec, typeInfo, ct)
            return toResult result
        }

    /// `bytes`, task-shaped.
    let bytesTask (spec: HttpSpec) (ct: Threading.CancellationToken) (runner: HttpRunner) =
        task {
            let! result = runner.BytesAsync(spec, ct)
            return toResult result
        }

    /// `text`, task-shaped.
    let textTask (spec: HttpSpec) (ct: Threading.CancellationToken) (runner: HttpRunner) =
        task {
            let! result = runner.TextAsync(spec, ct)
            return toResult result
        }
